### PR TITLE
[CELEBORN-1006] Add support for Apache Hadoop 2.x in Celeborn build

### PR DESCRIPTION
**Pull request by mridulm**
### What changes were proposed in this pull request?

Add support for Apache Hadoop 2.x in Celeborn build
Developers need to only specify their `hadoop.version`, and the build will pick the right profile internally based on the version to add the relevant dependencies.

### Why are the changes needed?

[hadoop-client-api](https://mvnrepository.com/artifact/org.apache.hadoop/hadoop-client-api) and [hadoop-client-runtime](https://mvnrepository.com/artifact/org.apache.hadoop/hadoop-client-runtime) were introduced in hadoop 3.x, while hadoop 2.x had [hadoop-client](https://mvnrepository.com/artifact/org.apache.hadoop/hadoop-client)
Celeborn depends on the former, and so requires hadoop 3.x to build.

Apache Spark dropped support for Hadoop 2.x only in the recent v3.5 ([SPARK-42452](https://issues.apache.org/jira/browse/SPARK-42452)). Given this, we have case where deployments on supported platforms like Spark 3.4 and older running on 2.x hadoop, will need to pull in hadoop 3.x just for Celeborn.

This PR uses `hadoop-client` when `hadoop.version` is specified as 2.x - and preserves existing behavior when `hadoop.version` is 3.x

Note - while using `hadoop-client` in 3.x is an option, hadoop community recommendation is to rely on `hadoop-client-api`/`hadoop-client-runtime`, hence making an effort to leverage that as much as possible.

### Does this PR introduce _any_ user-facing change?

Adds support for using 2.x for hadoop.version

### How was this patch tested?

Three combinations were tested:

* Default, without overriding hadoop.version

Dependencies:
```
$ build/mvn dependency:list 2>&1 | grep hadoop | sort | uniq
[INFO]    org.apache.hadoop:hadoop-client-api:jar:3.2.4:compile
[INFO]    org.apache.hadoop:hadoop-client-runtime:jar:3.2.4:compile
```

Will update this section again based on test suite results (which are ongoing)

* Setting hadoop.version to newer 3.3.0 explicitly

Dependencies:
```
$ ARGS="-Pspark-3.1 -Dhadoop.version=3.3.0" ; build/mvn dependency:list $ARGS 2>&1 | grep hadoop | sort | uniq
[INFO]    org.apache.hadoop:hadoop-client-api:jar:3.3.0:compile
[INFO]    org.apache.hadoop:hadoop-client-runtime:jar:3.3.0:compile
```

* Setting hadoop.version to older 2.10.0

Dependencies:
```
$ ARGS="-Pspark-3.1 -Dhadoop.version=2.10.0" ; build/mvn dependency:list $ARGS 2>&1 | grep hadoop | grep compile | sort | uniq
[INFO]    org.apache.hadoop:hadoop-auth:jar:2.10.0:compile -- module hadoop.auth (auto)
[INFO]    org.apache.hadoop:hadoop-client:jar:2.10.0:compile -- module hadoop.client (auto)
[INFO]    org.apache.hadoop:hadoop-common:jar:2.10.0:compile -- module hadoop.common (auto)
[INFO]    org.apache.hadoop:hadoop-hdfs-client:jar:2.10.0:compile -- module hadoop.hdfs.client (auto)
[INFO]    org.apache.hadoop:hadoop-mapreduce-client-app:jar:2.10.0:compile -- module hadoop.mapreduce.client.app (auto)
[INFO]    org.apache.hadoop:hadoop-mapreduce-client-common:jar:2.10.0:compile -- module hadoop.mapreduce.client.common (auto)
[INFO]    org.apache.hadoop:hadoop-mapreduce-client-core:jar:2.10.0:compile -- module hadoop.mapreduce.client.core (auto)
[INFO]    org.apache.hadoop:hadoop-mapreduce-client-jobclient:jar:2.10.0:compile
[INFO]    org.apache.hadoop:hadoop-mapreduce-client-shuffle:jar:2.10.0:compile -- module hadoop.mapreduce.client.shuffle (auto)
[INFO]    org.apache.hadoop:hadoop-yarn-api:jar:2.10.0:compile -- module hadoop.yarn.api (auto)
[INFO]    org.apache.hadoop:hadoop-yarn-common:jar:2.10.0:compile -- module hadoop.yarn.common (auto)
```

For each of the case above, build/test passes for each of the `ARGS`.

### CODE REVIEW

**Comment by mridulm** [comment]
+CC @otterc

File: client-mr/mr/pom.xml [hunk 1]
```diff
@@ -39,20 +39,48 @@
...
+      <id>hadoop-2</id>
+      <activation>
+        <property>
+          <name>hadoop-2-deps</name>
+        </property>
+      </activation>
+      <dependencies>
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>hadoop-client</artifactId>
```
Comment (waitinfuture) [inline client-mr/mr/pom.xml:80]: Seem we need to put this in LICENSE and LICENSE-binary, cc @pan3793 @FMX

Comment (waitinfuture) [reply client-mr/mr/pom.xml:80]: Seems hadoop-client will introduce several dependencies that are not needed, for example `hadoop-yarn*, hadoop-mapreduce*`, maybe we should exclude those.

File: .mvn/extensions.xml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<extensions xmlns="http://maven.apache.org/EXTENSIONS/1.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (cfmcgrady) [inline .mvn/extensions.xml:1]: license header?

Comment (mridulm) [reply .mvn/extensions.xml:18]: Added, thanks for catching this !

File: client-mr/mr/pom.xml (hunk 1 above)
Comment (mridulm) [reply client-mr/mr/pom.xml:80]: Updated `pom.xml` to exclude transitive deps which are not required, and updated LICENSE-binary.
Had to add `hadoop-auth` as well - which was getting used from `UserGroupInformation` (for `PlatformName`).

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks!

**Review by FMX — APPROVED** [review]
LGTM. Thanks.

**Comment by waitinfuture** [comment]
Thanks! Merged to main.

**Comment by mridulm** [comment]
Thanks for the reviews @waitinfuture, @cfmcgrady and @FMX !

**Pull request closed without merging**